File: app/src/main/java/com/bitflaker/lucidsourcekit/main/binauralbeats/RecyclerViewAdapterBackgroundNoisesManager.java
package com.bitflaker.lucidsourcekit.main.binauralbeats;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bitflaker.lucidsourcekit.data.BackgroundNoise;
import com.bitflaker.lucidsourcekit.databinding.EntryBinauralBackgroundNoiseBinding;

import java.util.List;

public class RecyclerViewAdapterBackgroundNoisesManager extends RecyclerView.Adapter<RecyclerViewAdapterBackgroundNoisesManager.MainViewHolderBackgroundNoises> {
    private final List<BackgroundNoise> backgroundNoises;
    private final Context context;
    private OnVolumeChanged mVolumeChangedListener;
    private OnEntryClicked mListener;

    public RecyclerViewAdapterBackgroundNoisesManager(Context context, List<BackgroundNoise> backgroundNoises) {
        this.context = context;
        this.backgroundNoises = backgroundNoises;
    }

    @NonNull
    @Override
    public MainViewHolderBackgroundNoises onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(context);
        return new MainViewHolderBackgroundNoises(EntryBinauralBackgroundNoiseBinding.inflate(inflater, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull MainViewHolderBackgroundNoises holder, int position) {
        holder.binding.txtBackgroundNoiseName.setText(backgroundNoises.get(position).getName());
        holder.binding.sldVolumeBackgroundNoise.setValue(backgroundNoises.get(position).getVolume());
        holder.binding.imgNoiseIcon.setImageDrawable(context.getResources().getDrawable(backgroundNoises.get(position).getIcon(), context.getTheme()));
        holder.binding.imgNoiseIcon.setOnClickListener(e -> {
            if(mListener != null) {
                // TODO display that it is paused by for example changing slider color to gray?
                backgroundNoises.get(position).setPaused(!backgroundNoises.get(position).isPaused());
                mListener.onEvent(backgroundNoises.get(position), position);
            }
        });
        holder.binding.sldVolumeBackgroundNoise.addOnChangeListener((slider, value, fromUser) -> {
            if(fromUser && mVolumeChangedListener != null){
                backgroundNoises.get(position).setVolume(value);
                mVolumeChangedListener.onEvent(backgroundNoises.get(position), position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return backgroundNoises.size();
    }

    public static class MainViewHolderBackgroundNoises extends RecyclerView.ViewHolder {
        EntryBinauralBackgroundNoiseBinding binding;

        public MainViewHolderBackgroundNoises(@NonNull EntryBinauralBackgroundNoiseBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public interface OnEntryClicked {
        void onEvent(BackgroundNoise noise, int position);
    }

    public void setOnEntryClickedListener(OnEntryClicked eventListener) {
        mListener = eventListener;
    }

    public interface OnVolumeChanged {
        void onEvent(BackgroundNoise noise, int position);
    }

    public void setOnVolumeChangedListener(OnVolumeChanged eventListener) {
        mVolumeChangedListener = eventListener;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/binauralbeats/RecyclerViewAdapterBackgroundNoisesManager.kt
package com.bitflaker.lucidsourcekit.main.binauralbeats

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView
import com.bitflaker.lucidsourcekit.databinding.EntryBinauralBackgroundNoiseBinding
import com.bitflaker.lucidsourcekit.main.binauralbeats.RecyclerViewAdapterBackgroundNoisesManager.MainViewHolderBackgroundNoises

class RecyclerViewAdapterBackgroundNoisesManager(
    private val context: Context,
    private val backgroundNoises: MutableList<BackgroundNoise>
) : RecyclerView.Adapter<MainViewHolderBackgroundNoises>() {
    class MainViewHolderBackgroundNoises(var binding: EntryBinauralBackgroundNoiseBinding) : RecyclerView.ViewHolder(binding.root)

    var onVolumeChangedListener: ((BackgroundNoise, Int) -> Unit)? = null
    var onEntryClickedListener: ((BackgroundNoise, Int) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainViewHolderBackgroundNoises {
        val binding = EntryBinauralBackgroundNoiseBinding.inflate(LayoutInflater.from(context), parent, false)
        return MainViewHolderBackgroundNoises(binding)
    }

    override fun onBindViewHolder(holder: MainViewHolderBackgroundNoises, position: Int) {
        holder.binding.txtBackgroundNoiseName.text = backgroundNoises[position].name
        holder.binding.sldVolumeBackgroundNoise.value = backgroundNoises[position].volume
        holder.binding.imgNoiseIcon.setImageDrawable(ResourcesCompat.getDrawable(context.resources, backgroundNoises[position].icon, context.theme))
        holder.binding.imgNoiseIcon.setOnClickListener {
            // TODO display that it is paused by for example changing slider color to gray?
            backgroundNoises[position].isPaused = !backgroundNoises[position].isPaused
            onEntryClickedListener?.invoke(backgroundNoises[position], position)
        }
        holder.binding.sldVolumeBackgroundNoise.addOnChangeListener { _, value, fromUser ->
            if (fromUser) {
                backgroundNoises[position].volume = value
                onVolumeChangedListener?.invoke(backgroundNoises[position], position)
            }
        }
    }

    override fun getItemCount(): Int {
        return backgroundNoises.size
    }
}
